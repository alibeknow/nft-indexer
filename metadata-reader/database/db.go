package database

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
	"go.mongodb.org/mongo-driver/mongo/writeconcern"
	"go.uber.org/zap"
)

type Config struct {
	Name             string
	Protocol         string
	Username         string
	Password         string
	Host             string
	Port             int
	SSLEnable        bool
	DirectConnection bool
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	if config.DirectConnection {
		url = fmt.Sprintf(
			"%s/?connect=direct",
			url,
		)
	}
	return url
}

type DB struct {
	log    *zap.Logger
	client *mongo.Client
	name   string
}

func Open(log *zap.Logger, ctx context.Context, config Config) (*DB, error) {
	opts := options.Client().ApplyURI(config.URL())

	if config.SSLEnable {
		path, _ := os.Getwd()
		pemPath := fmt.Sprintf("%s/../rds-combined-ca-bundle.pem", path)
		rootPEM, err := os.ReadFile(pemPath)
		if err != nil {
			return nil, err
		}

		roots := x509.NewCertPool()
		if ok := roots.AppendCertsFromPEM(rootPEM); !ok {
			return nil, fmt.Errorf("get certs from %s fail!\n", pemPath)
		}
		tlsConfig := &tls.Config{
			RootCAs:            roots,
			InsecureSkipVerify: true,
		}

		opts.SetReadPreference(readpref.Secondary())
		opts.SetWriteConcern(writeconcern.New(writeconcern.WMajority(), writeconcern.J(true), writeconcern.WTimeout(1000)))
		opts.SetTLSConfig(tlsConfig)
	}

	client, err := mongo.NewClient(opts)
	if err != nil {
		return nil, err
	}

	err = client.Connect(ctx)
	if err != nil {
		return nil, err
	}

	pingCtx, pingCancel := context.WithTimeout(ctx, 5*time.Second)
	defer pingCancel()
	err = client.Ping(pingCtx, nil)
	if err != nil {
		return nil, err
	}
	log.Info("Connected to DB", zap.String("url", config.URL()))

	return &DB{
		log:    log,
		client: client,
		name:   config.Name,
	}, nil
}

func (db *DB) Close() error {
	return db.client.Disconnect(context.Background())
}

func (db *DB) Metadata() *MetadataDB {
	coll := db.client.Database(db.name).Collection("metadata")
	return &MetadataDB{
		collection: coll,
	}
}
