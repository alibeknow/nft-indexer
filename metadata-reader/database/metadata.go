package database

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"metadata-reader/metadata"
)

type MetadataDB struct {
	collection *mongo.Collection
}

func (mDB MetadataDB) InsertOrUpdate(ctx context.Context, m metadata.Metadata) error {
	currentTime := time.Now()

	filter := bson.M{"_id": m.ID}
	update := bson.M{
		"$set": bson.M{
			"metadata":   m.Metadata,
			"type":       m.Type,
			"updated_at": currentTime,
		},
		"$setOnInsert": bson.M{
			"created_at": currentTime,
		},
	}

	opts := options.Update().SetUpsert(true)
	_, err := mDB.collection.UpdateOne(ctx, filter, update, opts)
	return err
}
