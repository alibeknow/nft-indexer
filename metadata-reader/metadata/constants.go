package metadata

type Chain string

const (
	ETH     Chain = "eth"
	POLYGON Chain = "polygon"
)

func GetMappings() map[string]interface{} {
	properties := make(map[string]interface{})

	properties["name"] = map[string]interface{}{
		"type":  "text",
		"index": true,
	}

	properties["description"] = map[string]interface{}{
		"type":  "text",
		"index": true,
	}

	properties["attributes"] = map[string]interface{}{
		"type":    "object",
		"enabled": true,
		"dynamic": true,
		"properties": map[string]interface{}{
			"key": map[string]interface{}{
				"type": "text",
			},
			"value": map[string]interface{}{
				"type": "text",
			},
		},
	}

	properties["chain"] = map[string]interface{}{
		"type":  "keyword",
		"index": true,
	}

	properties["contractAddress"] = map[string]interface{}{
		"type":  "text",
		"index": true,
	}

	properties["metadata"] = map[string]interface{}{
		"type":    "object",
		"enabled": false,
		"dynamic": true,
	}

	return map[string]interface{}{
		"properties": properties,
	}
}
