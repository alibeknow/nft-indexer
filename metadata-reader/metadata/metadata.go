package metadata

import (
	"context"
	"time"
)

type Metadata struct {
	ID        string     `bson:"_id"`
	Metadata  string     `bson:"metadata"`
	Type      string     `bson:"type"`
	CreatedAt *time.Time `bson:"created_at,omitempty"`
	UpdatedAt *time.Time `bson:"updated_at,omitempty"`
}

type DB interface {
	InsertOrUpdate(ctx context.Context, m Metadata) error
}
