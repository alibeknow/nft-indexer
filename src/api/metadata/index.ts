export * from './metadata.controller';
export * from './metadata.module';
export * from './metadata.service';
