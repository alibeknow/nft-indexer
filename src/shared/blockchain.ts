export enum Blockchain {
  ETH = 'eth',
  POLYGON = 'polygon',
}
