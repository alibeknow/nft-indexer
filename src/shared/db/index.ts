export * from './constants';
export * from './db.class';
export * from './db.deprecated';
export * from './db.config';
export * from './db.module';
export * from './db.provider';
export * from './mongo-sort';

