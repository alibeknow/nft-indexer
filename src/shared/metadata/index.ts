export * from './metadata-repo-db';
export * from './metadata-repo-storage';
export * from './on-chain-reader';
export * from './ipfs-reader';
export * from './http-reader';
export * from './interfaces';
export * from './constants';
export * from './read-content-headers';
export * from './pinata-reader';
export * from './metadata-search';
