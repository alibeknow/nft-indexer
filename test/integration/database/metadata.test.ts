import { DBClass, IDatabaseConfig, getDatabaseConfig } from '@shared/db';
import { MetadataRepoDB, UriProtocol } from '@shared/metadata';
import { fixtureMetdata } from '../fixtures/fixtures.metadata';
import { Blockchain } from '@shared/blockchain';

jest.setTimeout(50000);

describe('Metadata repo', () => {
  let db: DBClass;
  beforeAll(async () => {
    const dbConfig: IDatabaseConfig = getDatabaseConfig();
    db = new DBClass({ ...dbConfig, db: `db-metadata-repo-test-${dbConfig.db}` });
    await db.open();
  });

  describe('[save] method', () => {
    it('should save a metadata document to MongoDB in metadata collection', async () => {
      const contractAddress = '0xc4545a0f49e1246bdba8601fd79412d240cfc6aa';
      const tokenId = '0xc4425a0f49e1246bdba8601fd79412d240cfc6aa';
      const id = `${Blockchain.ETH}:${contractAddress}:${tokenId}`;

      const metadataRepo = new MetadataRepoDB(db);
      const saveMetadataResult = await metadataRepo.save(id, JSON.stringify({
        image: 'https://gateway.pinata.cloud/ipfs/QmWHgFha8FLDf6jeLhGLCrHbVsprT32dtPziBektrZNsMm/7883.png',
        tokenId: tokenId,
        name: 'Superlative Mutant Ape #7883',
        attributes: [
          {
            trait_type: 'Background',
            value: 'M1 Universe',
          },
          {
            trait_type: 'Clothes',
            value: 'M1 Coat',
          },
          {
            trait_type: 'Eyes',
            value: 'M1 Yellow Suprised',
          },
          {
            trait_type: 'Fur',
            value: 'M1 Green',
          },
          {
            trait_type: 'Hat',
            value: 'M1 Cowboy Hat',
          },
          {
            trait_type: 'Mouth',
            value: 'M1 Jovial',
          },
        ],
      }), UriProtocol.HTTPS);

      const doc = await db.metadata().findOne({ _id: id });

      expect(doc).toMatchSnapshot();
      expect(saveMetadataResult).toMatchSnapshot();
    });
  });

  describe('[getWhereIdIn] method', () => {
    it('should retrieve metadata documents by ids', async () => {
      const metadataRepo = new MetadataRepoDB(db);
      const metadataItems = [
        {
          id: 'eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e',
          metadata: JSON.stringify({
            description: 'Born out of a bad batch of fermented ...',
            image: 'ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png',
          }),
          type: UriProtocol.IPFS,
        },
        {
          id: 'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b',
          metadata: JSON.stringify({
            name: 'Braindom #2',
            description: 'COUNTDOWN OVER. MINTING LIVE.',
            image: 'ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png',
          }),
          type: UriProtocol.HTTPS,
        },
        {
          id: 'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72',
          metadata: JSON.stringify({
            name: 'NIKE  APE #114',
            description: 'COUNTDOWN OVER. MINTING LIVE.',
            image: 'http://gateway.pinatas.cloud/ipfs/nikebanc/(18).jpg',
          }),
          type: UriProtocol.HTTPS,
        },
      ];

      await Promise.all(metadataItems.map(({ id, metadata, type }) => metadataRepo.save(id, metadata, type)));

      const results1 = await metadataRepo.getWhereIdIn([
        'eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e',
        'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72',
      ]);

      const results2 = await metadataRepo.getWhereIdIn([
        'eth:fake-contract-id:fake-token-id',
      ]);

      const results3 = await metadataRepo.getWhereIdIn([
        'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b',
        'eth:fake-contract-id:fake-token-id',
      ]);

      const results4 = await metadataRepo.getWhereIdIn([
        'eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e',
        'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b',
        'eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72',
      ]);

      expect(results1).toHaveLength(2);
      expect(results1).toMatchSnapshot();
      expect(results2).toHaveLength(0);
      expect(results2).toMatchSnapshot();
      expect(results3).toHaveLength(1);
      expect(results3).toMatchSnapshot();
      expect(results4).toHaveLength(3);
      expect(results4).toMatchSnapshot();
    });
  });

  describe('[getIdsWhereIdIn]', () => {
    it('should return ids of available metadata by ids', async () => {
      const metadataRepo = new MetadataRepoDB(db);
      await Promise.all(fixtureMetdata.map(({ _id, metadata, type }) => metadataRepo.save(_id, metadata, type)));

      const result1 = await metadataRepo.getIdsWhereIdIn([
        'eth:0xE1E484251ffFee048A839bc2d68C160BD8C82aBB:0x0441',
        'eth:0x8393E4fd19623660074dDd66576Ec071F2D91292:0x0340',
        'eth:0x0e28B28681B1C6b58561A97720F50176ECA387E5:0x2612',
      ]);

      const result2 = await metadataRepo.getIdsWhereIdIn([
        'eth:fake-contract-id:fake-token-id',
      ]);

      const result3 = await metadataRepo.getIdsWhereIdIn([
        'eth:0xE1E484251ffFee048A839bc2d68C160BD8C82aBB:0x0441',
        'eth:fake-contract-id:fake-token-id',
        'eth:0x8393E4fd19623660074dDd66576Ec071F2D91292:0x0340',
      ]);

      const result4 = await metadataRepo.getIdsWhereIdIn([
        'eth:0xE1E484251ffFee048A839bc2d68C160BD8C82aBB:0x0441',
        'eth:0x8393E4fd19623660074dDd66576Ec071F2D91292:0x0340',
        'eth:0x8393E4fd19623660074dDd66576Ec071F2D91292:0x0341',
        'eth:0x0e28B28681B1C6b58561A97720F50176ECA387E5:0x2612',
      ]);

      expect(result1).toMatchSnapshot();
      expect(result2).toMatchSnapshot();
      expect(result3).toMatchSnapshot();
      expect(result4).toMatchSnapshot();
    });
  });

  afterEach(async () => {
    await db.metadata().deleteMany({});
  });

  afterAll(async () => {
    await db.close(true);
  });
});
