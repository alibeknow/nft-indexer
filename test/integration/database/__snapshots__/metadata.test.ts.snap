// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metadata repo [getIdsWhereIdIn] should return ids of available metadata by ids 1`] = `
[
  {
    "_id": "eth:0x0e28B28681B1C6b58561A97720F50176ECA387E5:0x2612",
  },
  {
    "_id": "eth:0x8393E4fd19623660074dDd66576Ec071F2D91292:0x0340",
  },
  {
    "_id": "eth:0xE1E484251ffFee048A839bc2d68C160BD8C82aBB:0x0441",
  },
]
`;

exports[`Metadata repo [getIdsWhereIdIn] should return ids of available metadata by ids 2`] = `[]`;

exports[`Metadata repo [getIdsWhereIdIn] should return ids of available metadata by ids 3`] = `
[
  {
    "_id": "eth:0x8393E4fd19623660074dDd66576Ec071F2D91292:0x0340",
  },
  {
    "_id": "eth:0xE1E484251ffFee048A839bc2d68C160BD8C82aBB:0x0441",
  },
]
`;

exports[`Metadata repo [getIdsWhereIdIn] should return ids of available metadata by ids 4`] = `
[
  {
    "_id": "eth:0x0e28B28681B1C6b58561A97720F50176ECA387E5:0x2612",
  },
  {
    "_id": "eth:0x8393E4fd19623660074dDd66576Ec071F2D91292:0x0340",
  },
  {
    "_id": "eth:0x8393E4fd19623660074dDd66576Ec071F2D91292:0x0341",
  },
  {
    "_id": "eth:0xE1E484251ffFee048A839bc2d68C160BD8C82aBB:0x0441",
  },
]
`;

exports[`Metadata repo [getWhereIdIn] method should retrieve metadata documents by ids 1`] = `
[
  {
    "_id": "eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72",
    "metadata": "{"name":"NIKE  APE #114","description":"COUNTDOWN OVER. MINTING LIVE.","image":"http://gateway.pinatas.cloud/ipfs/nikebanc/(18).jpg"}",
    "type": "https",
  },
  {
    "_id": "eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e",
    "metadata": "{"description":"Born out of a bad batch of fermented ...","image":"ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png"}",
    "type": "ipfs",
  },
]
`;

exports[`Metadata repo [getWhereIdIn] method should retrieve metadata documents by ids 2`] = `[]`;

exports[`Metadata repo [getWhereIdIn] method should retrieve metadata documents by ids 3`] = `
[
  {
    "_id": "eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b",
    "metadata": "{"name":"Braindom #2","description":"COUNTDOWN OVER. MINTING LIVE.","image":"ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png"}",
    "type": "https",
  },
]
`;

exports[`Metadata repo [getWhereIdIn] method should retrieve metadata documents by ids 4`] = `
[
  {
    "_id": "eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x5b",
    "metadata": "{"name":"Braindom #2","description":"COUNTDOWN OVER. MINTING LIVE.","image":"ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png"}",
    "type": "https",
  },
  {
    "_id": "eth:0x000386E3F7559d9B6a2F5c46B4aD1A9587D59Dc3:0x72",
    "metadata": "{"name":"NIKE  APE #114","description":"COUNTDOWN OVER. MINTING LIVE.","image":"http://gateway.pinatas.cloud/ipfs/nikebanc/(18).jpg"}",
    "type": "https",
  },
  {
    "_id": "eth:0x74ba9caaa5847e142263d5F7645fFeD92e5087E5:0x017e",
    "metadata": "{"description":"Born out of a bad batch of fermented ...","image":"ipfs://QmbCmZ2yLMG1U24kAo1or4Kn7w6tZ3u7KEZnC3s5Jo1wQL/382.png"}",
    "type": "ipfs",
  },
]
`;

exports[`Metadata repo [save] method should save a metadata document to MongoDB in metadata collection 1`] = `
{
  "_id": "eth:0xc4545a0f49e1246bdba8601fd79412d240cfc6aa:0xc4425a0f49e1246bdba8601fd79412d240cfc6aa",
  "metadata": "{"image":"https://gateway.pinata.cloud/ipfs/QmWHgFha8FLDf6jeLhGLCrHbVsprT32dtPziBektrZNsMm/7883.png","tokenId":"0xc4425a0f49e1246bdba8601fd79412d240cfc6aa","name":"Superlative Mutant Ape #7883","attributes":[{"trait_type":"Background","value":"M1 Universe"},{"trait_type":"Clothes","value":"M1 Coat"},{"trait_type":"Eyes","value":"M1 Yellow Suprised"},{"trait_type":"Fur","value":"M1 Green"},{"trait_type":"Hat","value":"M1 Cowboy Hat"},{"trait_type":"Mouth","value":"M1 Jovial"}]}",
  "type": "https",
}
`;

exports[`Metadata repo [save] method should save a metadata document to MongoDB in metadata collection 2`] = `
{
  "_id": "eth:0xc4545a0f49e1246bdba8601fd79412d240cfc6aa:0xc4425a0f49e1246bdba8601fd79412d240cfc6aa",
  "metadata": "{"image":"https://gateway.pinata.cloud/ipfs/QmWHgFha8FLDf6jeLhGLCrHbVsprT32dtPziBektrZNsMm/7883.png","tokenId":"0xc4425a0f49e1246bdba8601fd79412d240cfc6aa","name":"Superlative Mutant Ape #7883","attributes":[{"trait_type":"Background","value":"M1 Universe"},{"trait_type":"Clothes","value":"M1 Coat"},{"trait_type":"Eyes","value":"M1 Yellow Suprised"},{"trait_type":"Fur","value":"M1 Green"},{"trait_type":"Hat","value":"M1 Cowboy Hat"},{"trait_type":"Mouth","value":"M1 Jovial"}]}",
  "type": "https",
}
`;
